### PR TITLE
hsmd: signer with splicing onchain checks for same splicing/dualfunding operations

### PR DESCRIPTION
**Pull request by vincenzopalazzo**
>In general, a validating signer may be under a different operational environment than the node, and therefore may have a different source of on-chain data. The signer may therefore temporarily disagree on whether a funding or splice transaction is locked (buried). We would like to ensure agreement between the signer and the node on how to progress a channel's state.

For more information about the problem see https://github.com/ElementsProject/lightning/issues/6722

For the solution description see the commits

Fixes https://github.com/ElementsProject/lightning/issues/6722

### CODE REVIEW

**Comment by devrandom** [comment]
if I read the code correctly, this doesn't follow the design in #6722 - it calls both the check and the lock in a single execution.  the design intended to do both the check and the lock async and in different times in the negotiation flow.

I haven't evaluated the behavior of the code, just wanted to raise the difference from the design on the initial review pass.

**Comment by devrandom** [comment]
this said, this approach is likely OK for an initial version, since the likelihood of needing to block should be relatively low.

**Comment by ksedgwic** [comment]
This looks good to me

**Comment by vincenzopalazzo** [comment]
>if I read the code correctly, this doesn't follow the design in https://github.com/ElementsProject/lightning/issues/6722 - it calls both the check and the lock in a single execution. the design intended to do both the check and the lock async and in different times in the negotiation flow.

While re-reading the code I found this async execution problem too, this can happen a low with VLS or any remote signer that has a different source. I really think that we should change it to a timer base solution, but the effort is not trivial because the openingd do not have any concepts of timers so we should implement it somehow.

Before do this effort I would like to get some more review on the code.

**Review by ksedgwic — APPROVED** [review]

File: hsmd/libhsmd.c [hunk 1]
```diff
@@ -2021,6 +2065,8 @@ u8 *hsmd_init(struct secret hsm_secret,
 		WIRE_HSMD_SIGN_ANCHORSPEND,
 		WIRE_HSMD_SIGN_HTLC_TX_MINGLE,
 		WIRE_HSMD_SIGN_SPLICE_TX,
+		WIRE_HSMD_CHECK_OUTPOINT,
```
Comment (devrandom) [inline hsmd/libhsmd.c:2068]: I think we only want one capability for both messages.  we should have signers implement one but not the other.

Comment (ksedgwic) [reply hsmd/libhsmd.c:2068]: But `check_outpoint` is useful for other things than splices, no?

Comment (ksedgwic) [reply hsmd/libhsmd.c:2068]: Agree

Comment (vincenzopalazzo) [reply hsmd/libhsmd.c:2069]: I agree, makes sense!

Done :)

**Comment by nepet** [comment]
ACK 1fc383efd8a9b3a68b650f504dd3249f348c552d

**Review by rustyrussell — APPROVED** [review]
Ack 1fc383efd8a9b3a68b650f504dd3249f348c552d

It's simple, at least.  Probably better to refactor this function into common/ but it's OK as is.

**Comment by vincenzopalazzo** [comment]
> Probably better to refactor this function into common/ but it's OK as is.

Yeah, I agree I tried to write it but there was something not similar between the two functions. However, we iterate a lot over it so maybe now it is possible :)

I will take a look because keep just a function is also good for us

**Pull request merged**